### PR TITLE
hw03

### PR DESCRIPTION
**Pull request by Matveibos**

### CODE REVIEW

File: HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+# thi is class homework2
+class Homework3
+  SCAN_REGEX = %r{\d+\.\d\.\d+\.\d+\s-\s-\s\[\d+/\w+/\d+:\d+:\d+:\d+\s\+\d+\]\s"\w+\s/\S+\s\w+\S+\s\d+\s\S+\s\S+}
+  SPLIT_REGEX = /[\[(.*)\]]/
+
+  def task2(log)
+    res = []
```
Comment (alexshgov) [inline HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb:9]: Refactor without using the `res` temp variable

File: HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_03.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  def parsing_time(log)
+    check_required_lines(log).map { |string| Time.parse(string) }
+  end
+
+  def check_required_lines(log)
+    log.split("\n").select { |string| string != '' && string.include?('Calling core with action:') }
+  end
+
+  def result_array(array)
+    res = []
```
Comment (alexshgov) [inline HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_03.rb:24]: Refactor without using the `res` temp variable

File: HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_04.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+class Homework3
+  def task4(log)
+    res = {}
```
Comment (alexshgov) [inline HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_04.rb:5]: Refactor without using the `res` temp variable

File: HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_01.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class Homework3
+  def task1(log)
+    log.split("\n").select { |message| message.include?('error') }.first.to_s
```
Comment (B0RlS) [inline HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_01.rb:5]:
```suggestion
    log.split("\n").find { |message| message.include?('error') }.to_s
```
Could you read about the difference between `find` and `select` and write you answer in my telegram direct please?

File: HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+# thi is class homework2
+class Homework3
+  SCAN_REGEX = %r{\d+\.\d\.\d+\.\d+\s-\s-\s\[\d+/\w+/\d+:\d+:\d+:\d+\s\+\d+\]\s"\w+\s/\S+\s\w+\S+\s\d+\s\S+\s\S+}
```
Comment (B0RlS) [inline HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb:5]:
```suggestion
  SCAN_REGEX = %r{\d+\.\d\.\d+\.\d+\s-\s-\s\[\d+/\w+/\d+:\d+:\d+:\d+\s\+\d+\]\s"\w+\s/\S+\s\w+\S+\s\d+\s\S+\s\S+}.freeze
```

File: HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  SPLIT_REGEX = /[\[(.*)\]]/
+
+  def task2(log)
+    log.split("\n").delete_if { |string| good_format?(string) != string }.map { |str| print_string(str) }
+  end
+
+  private
+
+  def good_format?(string)
```
Comment (B0RlS) [inline HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb:14]:
```suggestion
  def valid_format?(string)
```

File: HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    string.scan(SCAN_REGEX).join
+  end
+
+  def print_string(string)
+    "DATE: #{string.split(SPLIT_REGEX)[4]} FROM: #{string.split(' ')[0]} TO: #{string.split(' ')[6].upcase}"
```
Comment (B0RlS) [inline HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb:19]: I would write a regular expression for address, ip and time and use them to select data from selected rows
up to u

File: HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_04.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+# this is class Homework3
+class Homework3
+  LET_REGEX = /[a-z]/
+  DIGITS_REGEX = /\d/
+  def task4(log)
+    { letters: log.downcase.scan(LET_REGEX).count, digits: log.scan(DIGITS_REGEX).count }
+  end
```
Comment (B0RlS) [inline HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_04.rb:9]:
```suggestion
  letters = string.count('a-zA-Z')
  digits = string.count('0-9')
  count_characters = { letters: letters, digits: digits }
```

File: HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb (hunk 5 above)
Comment (Matveibos) [reply HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb:5]: Снимок экрана от 2022-03-12 16-02-27

File: HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_01.rb (hunk 4 above)
Comment (Matveibos) [reply HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_01.rb:5]: Ok, fixed, thank you!

File: HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb (hunk 6 above)
Comment (Matveibos) [reply HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb:14]: Fixed, thanks!

File: HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_04.rb (hunk 8 above)
Comment (Matveibos) [reply HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_04.rb:9]: Thank you!

File: HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb (hunk 7 above)
Comment (Matveibos) [reply HW03/matsvei_basiayeu/matsvei_basiayeu_hw_03_t_02.rb:19]: I fixed it. Is that right?

REGEX_DATE = %r{\d+/\w+/\d+\:\d+\:\d+\:\d+\s\+\d+}
  REGEX_IP = /\d+\.\d+\.\d+\.\d+/
  REGEX_TEXT = %r{\w+/\d/\w+\s\w+/\d\.\d}


"DATE: #{string.match(REGEX_DATE)} FROM: #{string.match(REGEX_IP)} TO: #{string.match(REGEX_TEXT).to_s.upcase}"

**Pull request merged**